### PR TITLE
MH-12885 Capture Agent Access Management

### PR DESCRIPTION
**Pull request by krinnewitz**
This patch adds the possibility to configure which user (groups) have
access to which capture agents. Whenever a new capture agent is
registered, a new role is generated which a user needs to have when
changing scheduling information affecting the capture agent.

This includes:

- Editing schedule information of scheduled events
- Changing the workflow settings of scheduled events
- Scheduling new events
- Deleting scheduled events
- Deleting capture agents

The administrator as well as the organization administrator always has
access to all capture agents. For other users, access can be configured
by adding/removing the generated roles using the admin UI.

Here some user stories describing the reasoning behind this feature:

- As person responsible for scheduling, I do not want other users to mess up with capture agents I'm responsible for
- As a user, I want to be able to manage my recordings, including cutting of the video
- As a user, I want to see upcoming recordings ("scheduled events"), so that I know what will be recorded for me in case other persons are responsible for the actual recording
- As a institution, I might have several independent groups of people that are responsible for different sets of capture agents

Visually, this PR does not change much, but here are some impressions:

There are know three possibilities considering adding scheduled events:

1. The logged in user is system administrator (ROLE_ADMIN), tenant administrator (<Tenant Admin Role>) or has permissions to access all capture agents: Opencast will behave exactly the same as those roles have access to all capture agents

2. The logged in user is neither system administrator nor tenant administrator. He has only access rights to a subset of capture agents: The Add Event wizard looks still the same but the user can only select capture agents he has permissions to access

3. The logged in user has access to no capture agents at all: In that case, the Add Event Wizard has been adopted as this is the most common kind of users in Opencast setups that work with unprivileged users:

screen shot 2018-06-11 at 12 23 39

Users with no access to capture agents can see scheduled events based on the access policy, but they are not allowed to edit technical scheduling information:

screen shot 2018-06-11 at 12 23 57

This work is sponsored by SWITCH.

### CODE REVIEW

**Comment by krinnewitz** [comment]
I will rebase & extend this PR shortly since #231 is now merged

**Comment by krinnewitz** [comment]
Done

File: modules/admin-ui/src/main/resources/public/org/opencastproject/adminui/languages/lang-en_US.json
Comment (staubesv) [inline modules/admin-ui/src/main/resources/public/org/opencastproject/adminui/languages/lang-en_US.json]: Typo: per**s**mission

Comment (krinnewitz) [reply modules/admin-ui/src/main/resources/public/org/opencastproject/adminui/languages/lang-en_US.json]: thanks, fixed

**Comment by staubesv** [comment]
Hi @gregorydlogan 

Thanks for taking the review! 

I've realised that this PR is currently missing the documentation and created a [PR](https://github.com/plapadoo/opencast/pull/18) for that.

@krinnewitz Can you please merge that PR?

**Comment by gregorydlogan** [comment]
I haven't had a chance to actually look at this code yet, but ideally the docs and code would be in the same PR.  Can we rebase the docs into this PR, or is that going to cause issues on your end?

Edit: Ah, that's not the opencast repo, so I can't merge it in locally :)  Nevermind!

**Comment by staubesv** [comment]
@gregorydlogan The PR will be merged into this one. Currently waiting for @krinnewitz or @pmiddend to merge as I don't have write permissions for that repo, too.

**Comment by staubesv** [comment]
Argh... damn!

@krinnewitz I've provided some fixes to the community which once again break one of your pull requests... sorry for that :-)

Can you please fix the merge conflicts?

**Comment by krinnewitz** [comment]
ok, done @staubesv

**Comment by gregorydlogan** [comment]
@krinnewitz, more conflicts.

File: docs/guides/user/docs/advanced/capture-agent-access.md
Comment (gregorydlogan) [inline docs/guides/user/docs/advanced/capture-agent-access.md]: IIRC neither spaces nor quotes are allowed, but I suppose I probably forgot to add enforcement to that.  Sigh...

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java
Comment (gregorydlogan) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java]: If we're getting rid of the NotFoundException we should also remove line 339.

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java
Comment (gregorydlogan) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:123]: I see SecurityUtil.checkAgentAccess, and checkAgentAccess in a bunch of places.  IMO, either always wrap, or never wrap the check.

File: modules/admin-ui/src/main/webapp/scripts/modules/events/partials/eventActionsCell.html
Comment (gregorydlogan) [inline modules/admin-ui/src/main/webapp/scripts/modules/events/partials/eventActionsCell.html]: Auf Englisch bitte :)

File: modules/index-service/src/main/java/org/opencastproject/index/service/api/IndexService.java
Comment (gregorydlogan) [inline modules/index-service/src/main/java/org/opencastproject/index/service/api/IndexService.java]: Under what circumstances would we *not* want to check the access here?

**Review by gregorydlogan — CHANGES_REQUESTED** [review]
Looks good to me, will test tomorrow hopefully.

**Comment by pmiddend** [comment]
I fixed the conflicts. Thanks @gregorydlogan for your review so far. plapadoo is on vacation this week, so there won't be any more fixes till Friday.

File: docs/guides/user/docs/advanced/capture-agent-access.md
Comment (krinnewitz) [reply docs/guides/user/docs/advanced/capture-agent-access.md]: I will just change the example to "OpencastsCaptureAgent"

File: modules/admin-ui/src/main/webapp/scripts/modules/events/partials/eventActionsCell.html
Comment (krinnewitz) [reply modules/admin-ui/src/main/webapp/scripts/modules/events/partials/eventActionsCell.html]: whoops, looks like I forgot to delete this comment :P

File: modules/index-service/src/main/java/org/opencastproject/index/service/api/IndexService.java
Comment (krinnewitz) [reply modules/index-service/src/main/java/org/opencastproject/index/service/api/IndexService.java]: Seems to not be used. Removed.

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java
Comment (krinnewitz) [reply modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:123]: Done

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java
Comment (krinnewitz) [reply modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java]: Done

**Comment by krinnewitz** [comment]
@gregorydlogan thanks for your review so far. I have resolved the issues you came up with. Feel free to continue the review :-)

**Review by gregorydlogan — APPROVED** [review]

**Comment by gregorydlogan** [comment]
Looks great guys.  Sole complaint is that the CA is still visible there's no obvious reason why I couldn't schedule against it as a user who does not have permissions to use it.  Is this by design?  Either way, this seems like a reasonable PR to me!

**Comment by krinnewitz** [comment]
Thanks @gregorydlogan. You mean the CA's you don't have write access to shouldn't show up in the "locations" area of the adminUI? I have no strong feelings about that. Maybe you still want to see that the CA exists and which status it has?

**Comment by gregorydlogan** [comment]
@krinnewitz Yeah, fair enough.  This would be low impact even if the users aren't supposed to be able to see the agents, so let's see if we have complaints before we go to the work of fixing that :)

**Pull request merged**
